components/net/http_loader.rs

```diff
@@ -1887,31 +1902,179 @@ async fn http_network_fetch(
         let _ = fetch_terminated_sender.send(false);
     }
 
-    let response_future = obtain_response(
-        &context.state.client,
-        &url,
-        &request.method,
-        &mut request.headers,
-        body,
-        request
-            .body
-            .as_ref()
-            .map(|body| body.source_is_null())
-            .unwrap_or(false),
-        &request.pipeline_id,
-        request_id.as_deref(),
-        is_xhr,
-        context,
-        fetch_terminated_sender,
-    );
+    let (res, msg) = match &request.mode {
+        RequestMode::WebSocket { protocols } => {
+            // https://fetch.spec.whatwg.org/#websocket-opening-handshake
 
-    let pipeline_id = request.pipeline_id;
-    // This will only get the headers, the body is read later
-    let (res, msg) = match response_future.await {
-        Ok(wrapped_response) => wrapped_response,
-        Err(error) => return Response::network_error(error),
+            let (resource_event_sender, dom_action_receiver) = {
+                let mut websocket_chan = context.websocket_chan.as_ref().unwrap().lock().unwrap();
+                (websocket_chan.0.clone(),
+                 websocket_chan.1.take().unwrap())
+            };
+
+            // FIXME(pylbrecht): apply_hsts_rules()
+
+            let mut req_url = request.url();
+            if req_url.scheme() == "https" {
+                req_url.as_mut_url().set_scheme("wss").unwrap();
+            } else {
+                req_url.as_mut_url().set_scheme("ws").unwrap();
+            };
+            let req_origin = match request.origin {
+                Origin::Client => unreachable!(),
+                Origin::Origin(ref origin) => origin,
+            };
+
+            if should_request_be_blocked_due_to_a_bad_port(&req_url) {
+                let _ = resource_event_sender.send(WebSocketNetworkEvent::Fail);
+                // TODO(pylbrecht): not sure if we should return a network error here, but we
+                // have to early return something.
+                return Response::network_internal_error("port blocked");
+            }
```

Comment:
I think this is unnecessary duplication, as we call it in `main_fetch()` already.